### ISSUE TITLE
Sharee can edit (and delete) share-owner's public link shares

### ISSUE THREAD

**Issue by pascalwengerter** [owncloud/ocis#3498]
## Steps to reproduce

1. Go to https://ocis.ocis-web.latest.owncloud.works, log in as userA, create a folder with demo content, create public link for the folder and share folder with userB
2. Log in as other userB and accept share
3. As userB, navigate to your `Shares` and edit user public link for the folder shared by userA

## Expected behavior
Link shares from a user's personal space are not visible to sharees of said resources (for project spaces this might be a different thing, and where the bug is originally coming from). OC10 displays a different (and most likely correct) behaviour on https://demo.owncloud.com

## Actual behavior
Sharee can view, edit (and delete) share-owner's public link shares

**Comment by ScharfViktor** [comment]
Wow, I'm surprised. I can confirm that this is happening:
User B has "remove your share" and "change role" buttons available. 
- when User B deletes share - behavior looks like "decline"
-  when User B changes role, then - on the client it looks like he can really change it (of course it doesn't happen). 

![Screenshot 2022-04-14 at 11 58 31](https://user-images.githubusercontent.com/84779829/163363561-fc8f9881-a2ec-4737-aeac-b07d358d8362.png)

**Comment by C0rby** [comment]
I can't reproduce this. No matter what combination of permissions I use the sharee always sees this: 
![image](https://user-images.githubusercontent.com/5579653/170042868-b92d599e-2709-4d97-a1fb-58040b299e9f.png)

**Comment by C0rby** [comment]
Ah this is probably because we aren't using the `Shares` folder anymore.

**Comment by ScharfViktor** [comment]
Yes, I can confirm. User cannot edit owners public link

**Comment by ScharfViktor** [comment]
but user can see that link for folder is exist in "shared via link" section. Everything works weirdly there, maybe it's just in the development stage

https://user-images.githubusercontent.com/84779829/170059120-b2c87c6a-8429-435e-a01b-38f451cf75e4.mov

**Comment by C0rby** [comment]
So I have figured out that the problem is following:

When we are listing public shares we check if the user is either the share owner or creator. If neither is the case we stat the resource to check if the user has `ListGrants` permissions.
This was implemented to enable all space members to see shares.

But the same thing is happening when we have a user share from one user to another.

If einstein shares a file with richard, then richard has the `ListGrants` permission on the file.
This leads to the situation we are currently in.

The quick fix would be to somehow determine if the share is in a project space but in the long run we might want to implement a better permission handling.

**Comment by C0rby** [comment]
This could be a hacky quick fix:
```golang
if _, hit := cache[key]; !hit && !publicshare.IsCreatedByUser(local.PublicShare, u) {
    if _, spaceID := storagespace.SplitStorageID(local.ResourceId.StorageId); spaceID == local.Owner.OpaqueId {
        continue
    }
    sRes, err := client.Stat(ctx, &provider.StatRequest{Ref: &provider.Reference{ResourceId: local.ResourceId}})
```

But this only works until we introduce resharing, which happens soon so we need another solution...

**Comment by micbar** [comment]
@C0rby I understand that  ListingGrants is the reason why richard can see the public link.

But why does he then also have the "EditGrants"  permission?

**Comment by C0rby** [comment]
> But why does he then also have the "EditGrants" permission?

He doesn't. Maybe we aren't checking the permissions when updating or creating public links. I can't test it in the UI right now but I'll try to test it using curl.

**Comment by C0rby** [comment]
Ah, now I get the problem. Adding new shares as Richard doesn't work but editing or deleting the exisiting share does. I.e. updating the permissions of existing link shares.

Should we add a check there if the user has the `UpdateGrant` permission?

**Comment by micbar** [comment]
> Should we add a check there if the user has the UpdateGrant permission?

Yes i think so. Within the resharing implementation these things will become clearer. But technically it seems correct to distinguish between "List", "Add" and "update".
